helix-lsp/src/file_event.rs

```diff
@@ -0,0 +1,197 @@
+
+use globset::{GlobBuilder, GlobSetBuilder};
+use tokio::sync::mpsc;
+
+use crate::{lsp, Client};
+
+enum Event {
+    FileChanged {
+        path: PathBuf,
+    },
+    Register {
+        client_id: usize,
+        client: Weak<Client>,
+        registration_id: String,
+        ops: lsp::DidChangeWatchedFilesRegistrationOptions,
+    },
+    Unregister {
+        client_id: usize,
+        registration_id: String,
+    },
+    RemoveClient {
+        client_id: usize,
+    },
+}
+
+#[derive(Default)]
+struct ClientState {
+    client: Weak<Client>,
+    registered: HashMap<String, globset::GlobSet>,
+}
+
+/// The Handler uses a dedicated tokio task to respond to file change events by
+/// forwarding changes to LSPs that have registered for notifications with a
+/// matching glob.
+///
+/// When an LSP registers for the DidChangeWatchedFiles notification, the
+/// Handler is notified by sending the registration details in addition to a
+/// weak reference to the LSP client. This is done so that the Handler can have
+/// access to the client without preventing the client from being dropped if it
+/// is closed and the Handler isn't properly notified.
+#[derive(Clone, Debug)]
+pub struct Handler {
+    tx: mpsc::UnboundedSender<Event>,
+}
+
+impl Default for Handler {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Handler {
+    pub fn new() -> Self {
+        let (tx, rx) = mpsc::unbounded_channel();
+        tokio::spawn(Self::run(rx));
+        Self { tx }
+    }
+
+    pub fn register(
+        &self,
+        client_id: usize,
+        client: Weak<Client>,
+        registration_id: String,
+        ops: lsp::DidChangeWatchedFilesRegistrationOptions,
+    ) {
+        let _ = self.tx.send(Event::Register {
+            client_id,
+            client,
+            registration_id,
+            ops,
+        });
+    }
+
+    pub fn unregister(&self, client_id: usize, registration_id: String) {
+        let _ = self.tx.send(Event::Unregister {
+            client_id,
+            registration_id,
+        });
+    }
+
+    pub fn file_changed(&self, path: PathBuf) {
+        let _ = self.tx.send(Event::FileChanged { path });
+    }
+
+    pub fn remove_client(&self, client_id: usize) {
+        let _ = self.tx.send(Event::RemoveClient { client_id });
+    }
+
+    async fn run(mut rx: mpsc::UnboundedReceiver<Event>) {
+        let mut state: HashMap<usize, ClientState> = HashMap::new();
+        while let Some(event) = rx.recv().await {
+            match event {
+                Event::FileChanged { path } => {
+                    log::debug!("Received file event for {:?}", &path);
+
+                    let mut dropped_clients = Vec::new();
+                    for (id, client_state) in state.iter() {
+                        if client_state
+                            .registered
+                            .values()
+                            .any(|glob| glob.is_match(&path))
+                        {
```

Comment:
small nit instead of nesting everything inside an if you can simply continue if the condition doesn't match

Reply:
Makes sense, code updated